### PR TITLE
nbwidgets: add kymotracker widget

### PR DESCRIPTION
**Pull request by JoepVanlier**
**Why this PR?**
Using the algorithms by hand is really cumbersome. This widget helps users actually use it.

**Movies**
kymonew

**Where is algorithm 2?**
Algorithm 2 is pending.

**Why no widget coverage and where are the docs?**
I tried my best to test all the endpoints that actually do something for the different cases that can occur.
I have however not found a good way to test the actual widgets. In the last PR for this thing I will include both the documentation and an acceptance test-type document for how to test whether this thing isn't broken before a Pylake release.

Note: This PR depends on https://github.com/lumicks/pylake/pull/89 and https://github.com/lumicks/pylake/pull/90

### CODE REVIEW

File: lumicks/pylake/nb_widgets/tests/test_kymotracker_widget.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+from matplotlib.testing.decorators import cleanup
+import numpy as np
+import pytest
+
+
+@cleanup
+def test_widget_open(kymograph):
+    KymoWidgetGreedy(kymograph, 1, use_widgets=False)
+
+
```
Comment (mcsmurf) [inline lumicks/pylake/nb_widgets/tests/test_kymotracker_widget.py:11]: Great set of tests! For the future, I propose telling a story for each block/function...makes for much easier reviewing. Something like 
`# Given a kymograph,  alternating pixel thresholds should result in a kymo widget with corresponding lines`

**Review by mcsmurf — APPROVED** [review]
Looking good! Left a few thoughts here and there but nothing big. The tests were also a big help :rocket:

File: lumicks/pylake/kymotracker/kymoline.py [hunk 2]
```diff
@@ -191,6 +191,23 @@ def extend(self, other):
             raise TypeError(f"You can only extend a {self.__class__} with a {self.__class__} or "
                             f"{KymoLine}")
 
+    def remove_lines_in_rect(self, rect):
+        """Removes traces that fall in a particular region. Note that if any point on a line files inside the selected
```
[the remark this answers is not in the document]
Comment (JoepVanlier) [reply lumicks/pylake/kymotracker/kymoline.py:195]: 👍

File: lumicks/pylake/nb_widgets/kymotracker_widgets.py [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+        self.update_lines()
+
+    def _track(self, rect=None):
+        return filter_lines(self._algorithm(self.data, **self.algorithm_parameters, rect=rect), self.min_length)
+
+    def _connect_drag_callback(self):
+        def set_xlim(x, y, dx, dy):
+            old_xlims = np.array(self._axes.get_xlim())
+            self._dx = self._dx + dx
+            if abs(time.time() - self._last_update) > .033:
```
[the remark this answers is not in the document]
Comment (JoepVanlier) [reply lumicks/pylake/nb_widgets/kymotracker_widgets.py:82]: 30 frames per second. I added a comment to clarify this.
```# We don't need to update more than 30 times per second (1/30 = 0.033).```

File: lumicks/pylake/nb_widgets/kymotracker_widgets.py [hunk 4]
```diff
@@ -0,0 +1,360 @@
...
+        def set_value(value):
+            self.algorithm_parameters[name] = value
+
+        return ipywidgets.interactive(set_value,
+                                      value=slider_type(
+                                          description=description,
+                                          description_tooltip=tooltip,
+                                          min=minimum,
+                                          max=maximum,
+                                          step=step,
```
[the remark this answers is not in the document]
Comment (JoepVanlier) [reply lumicks/pylake/nb_widgets/kymotracker_widgets.py:141]: `step=step_size`

File: lumicks/pylake/nb_widgets/kymotracker_widgets.py [hunk 5]
```diff
@@ -0,0 +1,360 @@
...
+            show_lines=ipywidgets.ToggleButton(
+                description="Show Lines",
+                value=self.show_lines,
+                disabled=False,
+                button_style="",
+                tooltip="Show Lines\n\nDisabling this will hide all lines",
+            ),
+        )
+
+        all_button = ipywidgets.Button(description="Track all", tooltip="Reset all lines and track all")
```
[the remark this answers is not in the document]
Comment (JoepVanlier) [reply lumicks/pylake/nb_widgets/kymotracker_widgets.py:200]: 👍

[the remark this answers is not in the document]
Comment (JoepVanlier) [reply lumicks/pylake/nb_widgets/kymotracker_widgets.py:200]: 👍

File: lumicks/pylake/nb_widgets/tests/test_kymotracker_widget.py (hunk 1 above)
Comment (JoepVanlier) [reply lumicks/pylake/nb_widgets/tests/test_kymotracker_widget.py:11]: Will do!

File: lumicks/pylake/nb_widgets/kymotracker_widgets.py [hunk 6]
```diff
@@ -0,0 +1,360 @@
...
+            Which delimiter to use in the csv file.
+        sampling_width : int or None
+            When supplied, this will sample the source image around the kymograph line and export
+            the summed intensity with the image. The value indicates the number of pixels in either direction
+            to sum over.
+        """
+        self.lines.save(filename, dt, dx, delimiter, sampling_width)
+
+    def _load_from_ui(self):
+        self.lines = import_kymolines_from_csv(self.output_filename, self.data)
```
[the remark this answers is not in the document]
Comment (JoepVanlier) [reply lumicks/pylake/nb_widgets/kymotracker_widgets.py:126]: I can catch missing files. Do you think I should add a label to the widget where errors can be displayed?

[the remark this answers is not in the document]
Comment (JoepVanlier) [reply lumicks/pylake/nb_widgets/kymotracker_widgets.py]: image

I added a little status label that shows exceptions (see image).

[the remark this answers is not in the document]
Comment (mcsmurf) [reply lumicks/pylake/nb_widgets/kymotracker_widgets.py]: This is perfect

**Pull request merged**